### PR TITLE
mds: add dump_inode command

### PR DESCRIPTION
**Pull request by taodd**
1 when the mds cache is really big. it's hard to dump all the cache
2 most of time, we only want to know a specific inode info.

### CODE REVIEW

File: src/mds/MDSRank.cc [hunk 1]
```diff
@@ -1936,6 +1936,14 @@ bool MDSRankDispatcher::handle_asok_command(
       return true;
     }
     command_export_dir(f, path, (mds_rank_t)rank);
+  } else if (command == "dump inode") {
```
Comment (batrick) [inline src/mds/MDSRank.cc]: This sounds useful but I would prefer this took an actual inode number rather than a path.

In any case, I think the `dump tree` command with a depth of 0 would be equivalent to your proposed `dump inode`, right?

Comment (taodd) [reply src/mds/MDSRank.cc]: great point 👍 , didn't notice that dump tree command with depth == 0...

Regarding to the implementation of dump tree command, it would still go through all the inodes in cache, when the cache is too large, i think it's too expensive to use "dump tree" just for one inode, since  commands would hold the mds_lock. and i think the user may want to use "dump inode" frequently to monitor that specific inode sometime.

 so, can i do this by implementing two commands:
dump_inode_by_path <path>
dump_inode_by_ino  <inode number>

Comment (batrick) [reply src/mds/MDSRank.cc]: I think `dump tree` only lists the given inode if you give depth `0` (or is it `1`?), right? The size of the cache wouldn't matter?

Comment (taodd) [reply src/mds/MDSRank.cc]: yes, it only list the given inode, but the algorithm is traversing all the inodes in inode_map no matter if it's already found or not.  for huge cache and frequent use scenario,  i'm worried it would hold the mds_lock too long to just dump one inode.

Comment (batrick) [reply src/mds/MDSRank.cc]: Okay... that `MDCache::dump_cache` code is a bit messy. After thinking about this more, I would like to put a band-aid on the `dump_tree` command where `MDCache::dump_cache` will specialize the case where `dump_root !- ""` and `depth == 0`. In that case, it just looks up the file as in your `MDCache::dump_inode` function but prints it out exactly as in `MDCache::dump_cache` (i.e. don't change the format of the output).

I also noticed that `MDCache::dump_cache` is dumping directories in a special way by dumping dirfrags whereas `CDir::dump` does not. Perhaps we can add a flags argument to `CDir::dump` similar to what you did for `CInode::dump`. So, `CDir::dump` would also dump its `dirs` array of dentries (but not by default). Simlarly, we would also have a flag for `CInode::dump` to dump its in-cache dirfrags.

In summary, we'd have the `dump_func` closure in `MDCache::dump_cache` just call `CInode::dump` with the correct flags after filtering the inodes by path.

**Comment by taodd** [comment]
your suggestions are great,  i will open another pr
for cleanup the dump_cache. for the dump tree command, maybe we should not rely on inefficient dump_cache, and implement it just by traverse the tree directly.

**Comment by batrick** [comment]
This needs rebased too.

**Comment by taodd** [comment]
I've opened a another PR: 
https://github.com/ceph/ceph/pull/20472

**Comment by taodd** [comment]
I think dump inode info by inode number is useful too.
when we debug the osd log, file name is showed by inode number.
and sometimes we really want to know which file is the inode number point to.
so, l will do that in this PR.

**Comment by batrick** [comment]
Needs a rebase.

**Comment by taodd** [comment]
@batrick I'm working on this now

**Comment by taodd** [comment]
@batrick  I've created a new PR:
 https://github.com/ceph/ceph/pull/22569

**Comment by batrick** [comment]
@taodd in the future, you can just do a force push to your branch and that would update this PR.

**Comment by taodd** [comment]
@batrick got you !
